In [2]:
# source text format . For example if below is the input sequence, we will feed the data to the model as given in the 
#next line. We are going to use davincicode.txt file as the input which has more than 100K sentenses
data = 'Jack and Jill went up the hill To fetch a pail of water Jack fell down and broke his crown'

## We will feed the model, Line by Line sequence of text and genrate the output 

            X,						y

        _, _, _, _, _, Jack, 				and


        _, _, _, _, Jack, and 				Jill

        _, _, _, Jack, and, Jill,			went

        _, _, Jack, and, Jill, went,		up

        _, Jack, and, Jill, went, up,		the

        Jack, and, Jill, went, up, the,		hill

In [3]:
from keras.preprocessing.text import Tokenizer

In [4]:
import numpy as np
import re

In [5]:
book_text=open('../input/davincicodetxt/davincicode.txt',encoding='UTF-8').read()

In [6]:
len(book_text)

842670

In [7]:
sequences=list()
for line in book_text.split('.'):
    sequences.append(line)

In [8]:
len(sequences)

12293

In [9]:
sequences[2]

''

In [10]:
def clean_str(string):
  """
  String cleaning before vectorization
  """
  try:    
    string = re.sub(r'^https?:\/\/<>.*[\r\n]*', '', string, flags=re.MULTILINE)
    string = re.sub(r"[^A-Za-z]", " ", string)         
    words = string.strip().lower().split()    
    words = [w for w in words if len(w)>=1]
    if len(words)>1:
        return " ".join(words)	
    else:
        return 'NA'
    
  except:
    return ""

In [11]:
cleaned_seq=list() 

In [12]:
for line in sequences:
    cleaned_seq.append(clean_str(line))
    

In [13]:
cleaned_seq[2]

'NA'

In [14]:
len(cleaned_seq)

12293

In [15]:
cleaned_seq2=list()
for line in cleaned_seq:
    if line!='NA':
        cleaned_seq2.append(line)



In [16]:
len(cleaned_seq2)

11241

In [17]:
cleaned_seq2

['the da vinci code dan brown for blythe',
 'more than ever',
 'acknowledgments first and foremost to my friend and editor jason kaufman for working so hard on this project and for truly understanding what this book is all about',
 'and to the incomparable heide lange tireless champion of the da vinci code agent extraordinaire and trusted friend',
 'i cannot fully express my gratitude to the exceptional team at doubleday for their generosity faith and superb guidance',
 'thank you especially to bill thomas and steve rubin who believed in this book from the start',
 'my thanks also to the initial core of early in house supporters headed by michael palgon suzanne herz janelle moburg jackie everly and adrienne sparks as well as to the talented people of doubleday s sales force',
 'for their generous assistance in the research of the book i would like to acknowledge the louvre museum the french ministry of culture project gutenberg bibliotheque nationale the gnostic society library the dep

In [18]:
token=Tokenizer()

In [19]:
token.fit_on_texts(cleaned_seq2)

In [20]:
token.word_index

{'the': 1,
 'of': 2,
 'to': 3,
 'a': 4,
 'and': 5,
 'in': 6,
 'was': 7,
 'he': 8,
 's': 9,
 'langdon': 10,
 'his': 11,
 'had': 12,
 'you': 13,
 'it': 14,
 'i': 15,
 'that': 16,
 'sophie': 17,
 'on': 18,
 'as': 19,
 'is': 20,
 'at': 21,
 'her': 22,
 'with': 23,
 'for': 24,
 'this': 25,
 'she': 26,
 'not': 27,
 'teabing': 28,
 'from': 29,
 'said': 30,
 'be': 31,
 'an': 32,
 'him': 33,
 'but': 34,
 'have': 35,
 'they': 36,
 'now': 37,
 'could': 38,
 'out': 39,
 'what': 40,
 'were': 41,
 'fache': 42,
 'would': 43,
 'my': 44,
 'been': 45,
 'no': 46,
 'all': 47,
 'by': 48,
 'we': 49,
 'your': 50,
 'if': 51,
 'one': 52,
 't': 53,
 'into': 54,
 'up': 55,
 'are': 56,
 'me': 57,
 'grail': 58,
 'back': 59,
 'silas': 60,
 'over': 61,
 'them': 62,
 'like': 63,
 'about': 64,
 'only': 65,
 'man': 66,
 'their': 67,
 'there': 68,
 'grandfather': 69,
 'so': 70,
 'when': 71,
 'felt': 72,
 'sauniere': 73,
 'do': 74,
 'down': 75,
 'more': 76,
 'church': 77,
 'who': 78,
 'through': 79,
 'will': 80,
 'eyes':

In [21]:
# determine the vocabulary size
vocab_size = len(token.word_index) + 1
print('Vocabulary Size: %d' % vocab_size)

Vocabulary Size: 11186


In [22]:
encoded=token.texts_to_sequences(cleaned_seq2)

In [23]:
encoded

[[1, 132, 135, 286, 4673, 2332, 24, 4674],
 [76, 99, 305],
 [6387,
  157,
  5,
  3683,
  3,
  44,
  570,
  5,
  1533,
  6388,
  6389,
  24,
  971,
  70,
  414,
  18,
  25,
  3684,
  5,
  24,
  1088,
  1851,
  40,
  25,
  432,
  20,
  47,
  64],
 [5,
  3,
  1,
  6390,
  6391,
  6392,
  6393,
  6394,
  2,
  1,
  132,
  135,
  286,
  193,
  6395,
  5,
  1534,
  570],
 [15,
  571,
  914,
  4675,
  44,
  3082,
  3,
  1,
  3083,
  1422,
  21,
  4676,
  24,
  67,
  6396,
  498,
  5,
  3084,
  3085],
 [664,
  13,
  1227,
  3,
  3685,
  6397,
  5,
  6398,
  6399,
  78,
  637,
  6,
  25,
  432,
  29,
  1,
  1089],
 [44,
  2646,
  308,
  3,
  1,
  2333,
  2647,
  2,
  484,
  6,
  294,
  6400,
  1090,
  48,
  6401,
  6402,
  6403,
  6404,
  6405,
  6406,
  6407,
  6408,
  5,
  6409,
  3086,
  19,
  161,
  19,
  3,
  1,
  3686,
  251,
  2,
  4676,
  9,
  2648,
  687],
 [24,
  67,
  3087,
  3687,
  6,
  1,
  1034,
  2,
  1,
  432,
  15,
  43,
  63,
  3,
  6410,
  1,
  188,
  601,
  1,
  202,
  2063,

In [24]:
len(encoded)

11241

# We will create line based sequences 

In [25]:
sequence_list=list()

In [26]:
for line in encoded:
    for i in range(1,len(line)):
        sequence_list.append(line[:i+1])
        
        

In [27]:
print('Total Sequences_list:',len(sequence_list))

Total Sequences_list: 131120


In [28]:
sequence_list

[[1, 132],
 [1, 132, 135],
 [1, 132, 135, 286],
 [1, 132, 135, 286, 4673],
 [1, 132, 135, 286, 4673, 2332],
 [1, 132, 135, 286, 4673, 2332, 24],
 [1, 132, 135, 286, 4673, 2332, 24, 4674],
 [76, 99],
 [76, 99, 305],
 [6387, 157],
 [6387, 157, 5],
 [6387, 157, 5, 3683],
 [6387, 157, 5, 3683, 3],
 [6387, 157, 5, 3683, 3, 44],
 [6387, 157, 5, 3683, 3, 44, 570],
 [6387, 157, 5, 3683, 3, 44, 570, 5],
 [6387, 157, 5, 3683, 3, 44, 570, 5, 1533],
 [6387, 157, 5, 3683, 3, 44, 570, 5, 1533, 6388],
 [6387, 157, 5, 3683, 3, 44, 570, 5, 1533, 6388, 6389],
 [6387, 157, 5, 3683, 3, 44, 570, 5, 1533, 6388, 6389, 24],
 [6387, 157, 5, 3683, 3, 44, 570, 5, 1533, 6388, 6389, 24, 971],
 [6387, 157, 5, 3683, 3, 44, 570, 5, 1533, 6388, 6389, 24, 971, 70],
 [6387, 157, 5, 3683, 3, 44, 570, 5, 1533, 6388, 6389, 24, 971, 70, 414],
 [6387, 157, 5, 3683, 3, 44, 570, 5, 1533, 6388, 6389, 24, 971, 70, 414, 18],
 [6387,
  157,
  5,
  3683,
  3,
  44,
  570,
  5,
  1533,
  6388,
  6389,
  24,
  971,
  70,
  414,
  18,

Next , we need to pad the sequence using pad_sequence() in keras. Before that we need to find out the maximum lengh of the sequence so that we can pad all the sequences in the same length
    
# Pad Sequence

In [29]:
max_length=max([len(seq) for seq in sequence_list])
print ('maximum sequence length is', max_length)

maximum sequence length is 90


In [30]:
from keras.preprocessing.sequence import pad_sequences

In [31]:
sequence_padded_list=pad_sequences(sequence_list,maxlen=max_length,padding='pre')

In [32]:
sequence_padded_list

array([[    0,     0,     0, ...,     0,     1,   132],
       [    0,     0,     0, ...,     1,   132,   135],
       [    0,     0,     0, ...,   132,   135,   286],
       ...,
       [    0,     0,     0, ..., 11184,    48,  4673],
       [    0,     0,     0, ...,    48,  4673,  2332],
       [    0,     0,     0, ...,  4673,  2332, 11185]], dtype=int32)

Next, we need to split the data to input and output elements

# Split the data to input and output

In [33]:

sequence_padded_list=np.array(sequence_padded_list)



In [34]:
sequence_padded_list.shape

(131120, 90)

In [35]:
# Split the input and output data
X,y= sequence_padded_list[:,:-1],sequence_padded_list[:,-1]

In [36]:
X

array([[    0,     0,     0, ...,     0,     0,     1],
       [    0,     0,     0, ...,     0,     1,   132],
       [    0,     0,     0, ...,     1,   132,   135],
       ...,
       [    0,     0,     0, ...,     0, 11184,    48],
       [    0,     0,     0, ..., 11184,    48,  4673],
       [    0,     0,     0, ...,    48,  4673,  2332]], dtype=int32)

In [37]:
X.shape

(131120, 89)

In [38]:
y

array([  132,   135,   286, ...,  4673,  2332, 11185], dtype=int32)

In [39]:
y.shape

(131120,)

In [40]:
from keras.utils import to_categorical


In [41]:
vocab_size

11186

In [42]:
y_cat=to_categorical(y,num_classes=vocab_size)

In [43]:
y_cat

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 1.]], dtype=float32)

In [44]:
y_cat.shape

(131120, 11186)

Now , we need to train the model with the 131120 training examples each with max_length length.er 

First, we will create an embedding layer of 3 dimension as usual. 

vocab_size,

10 dimesnion vector for each word, 

(max_length-1) sequence length. -1 , becuase we excluded the output y already


In [45]:
import tensorflow as tf

# detect and init the TPU
tpu = tf.distribute.cluster_resolver.TPUClusterResolver()
tf.config.experimental_connect_to_cluster(tpu)
tf.tpu.experimental.initialize_tpu_system(tpu)

# instantiate a distribution strategy
tpu_strategy = tf.distribute.experimental.TPUStrategy(tpu)



In [46]:
from keras.models import Sequential
from keras.layers import Dense, Dropout,Embedding,LSTM
from keras.optimizers import RMSprop

We are going to use stacked LSTM with an additional LSTM layer 

In [48]:
# instantiating the model in the strategy scope creates the model on the TPU
#with tpu_strategy.scope():
model=Sequential()
model.add(Embedding(vocab_size,40,input_length=max_length-1))
model.add(LSTM(300,return_sequences=True))
model.add(LSTM(200))
model.add(Dense(vocab_size,activation='softmax'))
optim=RMSprop(lr=0.07)
model.compile(loss='categorical_crossentropy',optimizer='adam',metrics=['accuracy'])


In [49]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_1 (Embedding)      (None, 89, 40)            447440    
_________________________________________________________________
lstm_1 (LSTM)                (None, 89, 300)           409200    
_________________________________________________________________
lstm_2 (LSTM)                (None, 200)               400800    
_________________________________________________________________
dense (Dense)                (None, 11186)             2248386   
Total params: 3,505,826
Trainable params: 3,505,826
Non-trainable params: 0
_________________________________________________________________


# Compile and fit the model

In [50]:
model.fit(X,y_cat,epochs=110,batch_size=100,verbose=2)

Epoch 1/110
1312/1312 - 42s - loss: 7.0838 - accuracy: 0.0729
Epoch 2/110
1312/1312 - 42s - loss: 6.4416 - accuracy: 0.0939
Epoch 3/110
1312/1312 - 42s - loss: 6.0726 - accuracy: 0.1127
Epoch 4/110
1312/1312 - 42s - loss: 5.8099 - accuracy: 0.1273
Epoch 5/110
1312/1312 - 42s - loss: 5.5862 - accuracy: 0.1391
Epoch 6/110
1312/1312 - 42s - loss: 5.3823 - accuracy: 0.1477
Epoch 7/110
1312/1312 - 42s - loss: 5.4724 - accuracy: 0.1428
Epoch 8/110
1312/1312 - 42s - loss: 5.4155 - accuracy: 0.1480
Epoch 9/110
1312/1312 - 42s - loss: 5.1108 - accuracy: 0.1605
Epoch 10/110
1312/1312 - 42s - loss: 4.9144 - accuracy: 0.1702
Epoch 11/110
1312/1312 - 42s - loss: 4.7382 - accuracy: 0.1793
Epoch 12/110
1312/1312 - 42s - loss: 4.5618 - accuracy: 0.1885
Epoch 13/110
1312/1312 - 42s - loss: 4.3892 - accuracy: 0.1993
Epoch 14/110
1312/1312 - 42s - loss: 4.2247 - accuracy: 0.2121
Epoch 15/110
1312/1312 - 42s - loss: 4.0676 - accuracy: 0.2258
Epoch 16/110
1312/1312 - 42s - loss: 3.9171 - accuracy: 0.2395
E

In [51]:
model.save('keras_next_word_model_StakckedLSTM.h5')

In [52]:
import keras

In [53]:
model1=keras.models.load_model('./keras_next_word_model_StakckedLSTM.h5')



In [59]:
print(generate_seq(model1,token,max_length-1,'pagan',10))

pagan pantheonically pagan teabing exclaimed himself over the lid beside her


Now, lets test the model. we will generate a sequence using the below methods

In [55]:
# Generate sequence from language model

def generate_seq(model,token,max_length,seed_text,n_words):
    in_text=seed_text
    
    # generate the fixed number of words given by n_words in the input to the function
    for _ in range(n_words):
        # encode the text as integeer
        encoded_text=token.texts_to_sequences([in_text])[0]
        #prepad same as we did before training
        encoded_text=pad_sequences([encoded_text],maxlen=max_length,padding='pre')
        #predict the probabilities of the word
        y_pred=model.predict_classes(encoded_text,verbose=0)
        
        #map the index to the word
        
        outword=''
        for word, index in token.word_index.items():
            if index == y_pred:
                out_word = word
                break
        # append to input
        in_text += ' ' + out_word
    
    return in_text